src/context.cpp

```diff
@@ -309,19 +309,22 @@ class context_impl_t : public context_t {
     remove(const std::string& name) override {
         const holder_t scoped(*m_log, {{"source", "core"}});
 
-        std::unique_ptr<tcp_actor_t> service;
-
-        m_services.apply([&](service_list_t& list) {
+        auto extract_service = [&](service_list_t& list) -> std::unique_ptr<tcp_actor_t> {
             auto it = std::find_if(list.begin(), list.end(), match{name});
-            if(it != list.end()) {
-                service = std::move(it->second);
-                list.erase(it);
-            } else {
-                throw cocaine::error_t("service '{}' doesn't exist", name);
+            if(it == list.end()) {
+                 return nullptr;
             }
-        });
+            auto service = std::move(it->second);
+            list.erase(it);
+            return service;
+        };
 
-        service->terminate();
+        auto service = m_services.apply(extract_service);
+        if (service) {
+            service->terminate();
+        } else if (!(service = extract_service(m_unpublished))) {
```

Comment:
Unpublished service should be terminated too, shouldn't it?

Reply:
Yeah, I was gonna do that.
https://github.com/interiorem/cocaine-core/pull/6#issuecomment-437189240

Termination is not required for unpublished actor and lead to crash, but it fixable and will be better for using. I agree.

Reply:
We decided to do it in separate task. I added TODO.